Track Processing update

tschuh opened the pull request:
This PR updates mainly the modules TrackerTFP, TrackFindingTracklet and TrackTrigger.
It  brings the Kalman Filter, Track Multiplexer, Duplicate Removal and Track Quality steps up to date (https://indico.cern.ch/event/1449862/contributions/6116147/attachments/2924526/5133676/thomas.pdf)

Detailed list for Ian:
1. DataFormats/L1TrackTrigger
  - TTBV update
  - added kf chi2 output product (pair of doubles)
2. L1Trigger/TrackFindingTracklet
  - HitPatternHelper updated to cope changes, likely broken
  - KFin deleted
  - KalmanFilter added
  - 5 parameter fit option added (https://indico.cern.ch/event/1449862/contributions/6116147/attachments/2924526/5133676/thomas.pdf)
  - ProducerKalmanFilter added
  - State added
  - DRin renamed in TrackMultiplexer
  - L1FPGATrackProducer updated
  - DR renamed in DuplicateRemoval and updated
  - KFout removed
  - TBout removed
  - ChannelAssignment updated
  - FitTrack updated
  - L1TrackNtupleMaker updated with layer encoding specifics
3. L1Trigger/TrackTrigger
  - SensorModule updated
  - Setup updated
  - L1TrackQuality moved to trackerTFP
4. L1Trigger/TrackerDTC
  - DTC updated
  - LayerEncoding updated
  - Stub updated
5. L1Trigger/TrackerTFP
  - L1TrackQuality  renamed in TrackQuality and turned into ESProduct
  - TrackFindingProcessor added.
  - CleanTrackBuilder added
  - DataFormats updated
  - Demonstrator updated
  - Duplicate Removal added
  - GeometricProcessor updated
  - HoughTransform updated
  - KalmanFilter updated
  - LayerEncoding updated
  - MiniHoughTransform deleted
  - State updated
  - ZHoughTransform deleted
  - KFin deleted

tomalin committed:
Run combined modules by default (#265)

* Make combined modules default

* tweak

* Improve USEHYBRID ifdef range

* Fix compiler error for pure Tracklet algo

* Move fitpattern.txt refs, so only used for pure Tracklet algo

* code format

cgsavard committed:
Bin TQ MVA output (#272)

* add TQ MVA output binning scheme

* remove code no longer used

* move logistic sigmoid into TTTrack class

* fix for proper precision after MVA conversion in track word

* label all MVA variables before logistic sigmoid with 'Pre'

* formatting

* fix filling of track word before running TQ MVA

* switch to post-sigmoid mva bins

* remove any onnx instances

* fix MVA1 initialization

* bin TQ MVA in pre-sigmoided bins in KFout

* move pre-sigmoid bins to track quality class

* change minimum bin

tomalin commented:
As you're changing 160 files, the text describing this PR needs to be longer and more detailed. What are the changes that have been implemented? And how do they affect tracking performance? Have you given a talk which describes in detail the changes and the effect on performance. If so, please link it from here.

tomalin commented:
This fails CI quite spectacularly https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/pipelines/7858190 . e.g. For HYBRID_NEWKF, the tracking efficiency is up by 0.3% (good), but the number of reconstructed tracks is down by 80% (very weird, given that no efficiency is lost!), and the z0 resolution is a factor 5 worse (bad).

For comparison, this is the git CI for a dummy PR that changed no code https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/pipelines/7924082 .

tomalin commented:
> This fails CI quite spectacularly https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/pipelines/7858190 . e.g. For HYBRID_NEWKF, the tracking efficiency is up by 0.3% (good), but the number of reconstructed tracks is down by 80% (very weird, given that no efficiency is lost!), and the z0 resolution is a factor 5 worse (bad).
> 
> For comparison, this is the git CI for a dummy PR that changed no code https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/pipelines/7924082 .

==== Thomas says CI failure explained by poor z0 resolution. He's retuning the digitisation ranges to fix this. But he believes that the tracklet helix parameters are wrong (inconsistent with seed positions by ~1cm) for a small fraction of the tracklets, which is also contributing to poor z0 resolution.

tschuh commented:
new KF maths are fixed now, also passing CI using new KF. However, CI using old KF fails due to slightly lower z0 resolution. My PR should not impact those...

tomalin commented:
I understand that the ProducerTM usually orders tracks according to their seed type. This is important for the functioning of the DR, so should be described in the comment at the top of the class header file, which explains what the class does. And at the line of the code that causes this ordering to happen, an additional comment should appear to highlight it. A comment mentioning this inside ProducerDR would also be wise.

tschuh commented:
> I understand that the ProducerTM usually orders tracks according to their seed type. This is important for the functioning of the DR, so should be described in the comment at the top of the class header file, which explains what the class does. And at the line of the code that causes this ordering to happen, an additional comment should appear to highlight it. A comment mentioning this inside ProducerDR would also be wise.

done

tomalin commented:
@Chriisbrown & @cgsavard -- Thomas's overview description of this PR says "L1TrackQuality renamed in TrackQuality and turned into ESProduct". Are you both and Andreas happy with this change?

tomalin commented on L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py:
```diff
@@ -52,7 +51,7 @@
     process.TTClustersFromPhase2TrackerDigis,
     process.TTStubsFromPhase2TrackerDigis,
-    process.TrackerDTCProducer,
+    process.ProducerDTC,
```
Given that every sub-detector in CMS has DTCs, is it reasonable to use the name ProducerDTC. which gives no clue which subdetector it's for?

tschuh replied:
as far as I know only IT and OT are calling their back end boards DTC. I am not aware of any other subsystem doing this.

tomalin commented:
At the top of the header of the KF class where you do the maths (KalmanFilter.h?), add the comments copied from L7-31 of https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTMTT/interface/KFParamsComb.h#L7 , so the maths is documented. Although as you treat r-phi and r-z planes independently, you may need to modify a few lines of this.

tomalin commented:
Overview description of this PR should mention that new class TrackFindingTracklet/ProducerKF has been added.

tomalin commented on L1Trigger/TrackFindingTracklet/interface/DuplicateRemoval.h:
```diff
@@ -9,42 +9,39 @@
+  /*! \class  trklet::DuplicateRemoval
    *  \brief  Class to bit- and clock-accurate emulate duplicate removal
    *          DR identifies duplicates based on pairs of tracks that share stubs in at least 3 layers.
    *          It keeps the first such track in each pair.
```
Modify this comment to make clear that the track order is determined by TrackMultiplexer.

tschuh replied:
done

tomalin replied:
Modify comment to mention ProducerTM, since it's not obvious that the Track Multiplexer can be found there.

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/src/KalmanFilter.cc:
```diff
@@ -0,0 +1,562 @@
+    conv(streamsStub, streamsTrack);
+  }
+
+  // calulcate seed parameter
```
Typo in this comment

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/src/KalmanFilter.cc:
```diff
@@ -0,0 +1,562 @@
+  }
+
+  // calulcate seed parameter
+  void KalmanFilter::calcSeeds() {
```
I guess KalmanFilter::calcSeeds() calculates the helix params & their cov. matrix from a pair of stubs. Make comment describing this function a bit longer to say so.
I assume the maths used is that described in the (current) section 3.4 of the new TMTT paper. But I can't recognise it in this function?

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/src/KalmanFilter.cc:
```diff
@@ -0,0 +1,562 @@
+    }
+  }
+
+  // apply final cuts
```
What does "final" cuts mean in this comment? Are these cuts applied at the end of the KF fit to a track, after all stubs have been added to a state. Or are they cuts applied after adding each stub?

tschuh replied:
the former

tomalin commented on L1Trigger/TrackFindingTracklet/interface/State.h:
```diff
@@ -0,0 +1,146 @@
+
+namespace trklet {
+
+  // Class to represent a Kalman Filter State
```
Change comment to "Kalman FIlter helix State" to be a bit less obscure.

cgsavard commented on L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc:
```diff
@@ -195,26 +194,30 @@ class L1FPGATrackProducer : public edm::one::EDProducer<edm::one::WatchRuns> {
+
+  // helper class to store Track Trigger configuration
+  const Setup* setup_ = nullptr;
+  // helper class to store Tracklet spezific configuration
```
typo in "specific"

tschuh replied:
done

cgsavard commented on L1Trigger/TrackFindingTracklet/src/HitPatternHelper.cc:
```diff
@@ -21,13 +21,13 @@ namespace hph {
         useNewKF_(iConfig.getParameter<bool>("useNewKF")),
         chosenRofZNewKF_(setupTT_.chosenRofZ()),
-        etaRegionsNewKF_(setupTT_.boundarieEta()),
+        //etaRegionsNewKF_(setupTT_.boundarieEta()),
```
Should we just remove this if it's commented out?

tschuh replied:
oki doki

cgsavard commented on L1Trigger/TrackFindingTracklet/src/TrackMultiplexer.cc:
```diff
@@ -0,0 +1,493 @@
+            stub->z_ -= dPhi / inv2R * cot;
+          }
+        }
+        // checl track validity
```
typo in "check"

tschuh replied:
done

cgsavard commented on L1Trigger/TrackerTFP/python/Demonstrator_cfi.py:
```diff
@@ -3,9 +3,14 @@
-  LabelOut = cms.string( "TrackerTFPProducerKF"            ), #
-  DirIPBB  = cms.string( "/heplnw039/tschuh/work/proj/kf/" ), # path to ipbb proj area
-  RunTime  = cms.double( 6.0 )                                # runtime in us
+  LabelIn  = cms.string( "roducerDR"               ), #
```
Is this and the next line missing the "P" in "Producer"?

tschuh replied:
done

cgsavard commented on L1Trigger/TrackerTFP/python/Producer_cfi.py:
```diff
@@ -2,23 +2,19 @@
+  InputLabelCTB    = cms.string( "ProducerHT"    ),  #
+  InputLabelKF     = cms.string( "ProducerCTB"   ),  #
+  InputLabelDR     = cms.string( "ProducerKF"    ),  #
+  InputLabelTQ     = cms.string( "ProducerKF"    ),  #
```
I thought the order was DR -> KF -> TQ -> TFP, is there a mistake here? Is the DR output being used anywhere?

tschuh replied:
done and that is the tmtt track reco chain which runs KF -> DR -> TQ

cgsavard replied:
I'm a bit confused about this other chain we have. Why is this in the TFP folder if it's only for TMTT? And is this actually ever used then?

tschuh replied:
I am not good in naming stuff. I put my dtc emulator in TrackerDTC and my tfp emulator in TrackerTFP.

cgsavard commented on L1Trigger/TrackerTFP/src/TrackFindingProcessor.cc:
```diff
@@ -0,0 +1,266 @@
+    else if (invR > -TTTrack_TrackWord::minRinv)
+      invR = -TTTrack_TrackWord::minRinv - df->format(Variable::inv2R, Process::dr).base();
+    const double chi20 = chi20_ / (hitPattern_.count() - 2);
+    const double chi21 = chi21_ / (hitPattern_.count() - 2);
```
Why did we switch the convention of chi2rphi and chi2rz to chi20 and chi21? I think this change only adds confusion.

tschuh replied:
sorry I am used the the kf convention, but I don't care how we call them, so I changed it.

cgsavard replied:
Thank you for the changes. There are still a couple of instances where chi20 and chi21 remain which I think is fine but can you include comments that this corresponds to chi2rphi and chi2rz, such as in L1Trigger/TrackFindingTracklet/interface/State.h and L1Trigger/TrackerTFP/interface/State.h and L1Trigger/TrackTrigger/python/Setup_cfi.py ?

tschuh replied:
I will stick to the kalman filter convention inside the kalman filter, state and its configuration.

cgsavard replied:
Can you add some comments to these variables to explain what they are? Specifically in L1Trigger/TrackFindingTracklet/interface/State.h and L1Trigger/TrackerTFP/interface/State.h I notice that the other variables have comments to describe them and not the chi2 variables.

cgsavard commented on L1Trigger/TrackerTFP/src/TrackFindingProcessor.cc:
```diff
@@ -0,0 +1,266 @@
+      static constexpr unsigned int nPar0 = 2;
+      static constexpr unsigned int nPar1 = 2;
+      const double aChi2xyfit = it->chi20_ / (it->hitPattern_.count() - nPar0);
+      const double aChi2zfit = it->chi21_ / (it->hitPattern_.count() - nPar1);
```
chi2rphi and chi2rz should not be pdof when initializing a tttrack. The pdof variations are calculated in the TTTrack class so this is an error.

tschuh replied:
good to know, done

cgsavard commented on L1Trigger/TrackerTFP/src/TrackQuality.cc:
```diff
@@ -0,0 +1,258 @@
+    const int ninterior = nstub - 1;
+    const int chi2B = tq->toBinChi2B(0);
+    const int chi20 = tq->toBinChi20(trackChi20);
+    const int chi21 = tq->toBinChi21(trackChi21);
```
chi2rphi and chi2rz need to be pdof before binning, this is an error

tschuh replied:
I do not want to change the behavior of the class. When this emulates the f/w correctly we would introduce a f/w to s/w mismatch. @Chriisbrown may tell us what the f/w does.

cgsavard replied:
I see. Given the chi2rz and chi2rphi were not binned pdof before this pr then I guess this is keeping it matching what it was before. However, this does mean the MVA inputs are wrong so I think we should change it both here and in the firmware to get matching again, but I will let Chris comment here.

cgsavard commented on L1Trigger/TrackerTFP/src/TrackQuality.cc:
```diff
@@ -0,0 +1,258 @@
+    const double zT = tq->scaleZT(df->format(Variable::zT, Process::kf).integer(track.zT()));
+    const int nstub = hitPattern.count();
+    const int ninterior = nstub - 1;
+    const int chi2B = tq->toBinChi2B(0);
```
We should fill this with the simulation value for the time being with a note that this is the case. Otherwise the miscalculation of the MVA will break many things down the chain.

tschuh replied:
done

cgsavard commented on L1Trigger/TrackerTFP/src/TrackQuality.cc:
```diff
@@ -0,0 +1,258 @@
+      trackChi21 = tq->format(VariableTQ::chi21).range() - 1e-11;
+    // calc bdt inputs
+    const double cot = tq->scaleCot(df->format(Variable::cot, Process::dr).integer(track.cot()));
+    const double zT = tq->scaleZT(df->format(Variable::zT, Process::kf).integer(track.zT()));
```
This should be the z0 variable that is used by the tttracks. This current input to the MVA is wrong. And it should be scaled. See the variable calculated in L226 below.

tschuh replied:
ah here I was already anticipating a new TQ tune using zT instead of z0 which we should do in future. But I replaced it now with z0 as requested.

cgsavard replied:
I will pass this information along to Andreas who will take over the retraining of the new TQ BDT

cgsavard commented on L1Trigger/TrackerTFP/src/TrackQuality.cc:
```diff
@@ -0,0 +1,258 @@
+    const double cot = tq->scaleCot(df->format(Variable::cot, Process::dr).integer(track.cot()));
+    const double zT = tq->scaleZT(df->format(Variable::zT, Process::kf).integer(track.zT()));
+    const int nstub = hitPattern.count();
+    const int ninterior = nstub - 1;
```
This variable is not calculated correctly. See lines 209-216 below, or you can also use the HitPatternHelper function and grab the calculated numMissingInterior1_ from the hitpattern

tschuh replied:
same as a couple of comments before, I don't want to change the behavior without knowing what the f/w expects. I would not use the HitPatternHelper, it is likely broken after this PR.

cgsavard replied:
The behavior has already been changed actually. Before this PR it was being calculated correctly, just in another way then I mentioned. So if you want to keep the firmware and software matching then it will need to be fixed. You can find the calculation before which Chris matched the firmware to [here](https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_0_0_pre2/L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc#L202-L231). I suggest using the simpler calculation which produces the same results and only builds off the hitpattern bits [here](https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_0_0_pre2/L1Trigger/TrackTrigger/src/L1TrackQuality.cc#L39-L46).

tschuh replied:
you are right, I overlooked a not. I fixed it.

cgsavard replied:
This calculation is still incorrect. The correct calculation using the TTBV class functions is ```hitPattern.count(hitPattern.plEncode(), hitPattern.pmEncode(), false)```. Can we also change the variable name to have "miss" in it? ninterior sounds like it should be counting the hits and not the misses. I suggest ```n_missint```.

tschuh replied:
The calculations match the previous code, I don't want to change the behavior. I understand that it is likely buggy. But I can do the name change.

cgsavard replied:
Yes ok I do see that the calculation you have now has matches what was done before. I will open up an issue marking that both this and chi2rz and chi2rphi are incorrectly calculated so that we do not forget.

cgsavard commented on L1Trigger/TrackerTFP/src/TrackFindingProcessor.cc:
```diff
@@ -0,0 +1,266 @@
+      static constexpr unsigned int nPar1 = 2;
+      const double aChi2xyfit = it->chi20_ / (it->hitPattern_.count() - nPar0);
+      const double aChi2zfit = it->chi21_ / (it->hitPattern_.count() - nPar1);
+      const double trkMVA1 = it->mva_;
```
This is currently a binned value but it should not be when setting the TTTrack. It should be the 0-1 probability value.

tschuh replied:
good to know, done

cgsavard commented on L1Trigger/TrackerTFP/src/TrackQuality.cc:
```diff
@@ -0,0 +1,258 @@
+      fillDataFormats<++v>(iConfig);
+  }
+
+  // TQ MVA bin conversion LUT
+  constexpr array<double, numBinsMVA_> TrackQuality::mvaPreSigBins() const {
+    array<double, numBinsMVA_> lut = {};
+    lut[0] = -16.;
+    for (int i = 1; i < numBinsMVA_; i++)
+      lut[i] = invSigmoid(TTTrack_TrackWord::tqMVABins[i]);
+    return lut;
+  }
```
There was a discussion in PR #272 about how to set things up such that we avoid ever having to do the inverse sigmoid. I believe this was mainly to better compare with the firmware as we will not do this conversion there. Since it was recently decided there, I think we should stay true to that.

cgsavard commented on L1Trigger/TrackerTFP/src/TrackQuality.cc:
```diff
@@ -0,0 +1,258 @@
+    conifer::BDT<ap_fixed<10, 5>, ap_fixed<10, 5>> bdt(tq->model().fullPath());
+    // collect features and classify using bdt
+    const vector<ap_fixed<10, 5>>& output = bdt.decision_function({cot, zT, chi2B, nstub, ninterior, chi20, chi21});
+    const float mva = output[0].to_float();
```
Once the above changes to the input values are fixed, it should be checked that the MVA distribution still has a peak of fake tracks at 0 and real tracks at 1 to make sure it's working again. Right not the variable is nonsense.

tschuh replied:
as long as it is not retrained to the current z0, cot and chi2s digitization we should not expect great performance. When we retrain we should change the input variables to zT(layer encoding granularity), chi2B, hitPattern, chi2rphi, chi2rz. And maybe create a version without chi2B to have something we can test in f/w.

cgsavard replied:
Sure I do think the performance won't be as good as before, but we should still expect a fake track peak about 0 and real track peak around 1 so it would be nice to double check this is the case and nothing else is way off. I agree on the retraining and will pass this info along.

cgsavard reviewed (commented):
I added the comments I could in the parts that I am familiar with. Unfortunately there are a lot of files I am unfamiliar with and so I couldn't be of much help there. I am not opposed to changing tracker quality to its own EDProducer if you feel that it is a better setup. Although right now the HYBRID version still uses it as a function and not EDProducer. Can we make that consistent?

There are three main concerns I brought up:
1. The change in naming scheme from chi2rz and chi2rphi to chi20 and chi21 only adds confusion to any others who look in. I think it's better to change these to the naming scheme that is generally accepted.
2. The MVA1 variable is not being calculated correctly as many of the BDT inputs are wrong. The correct variables that should be input can be seen further down in the TrackQuality.cc file and I made notes on changes need to each variable.
3. The tttracks being created in TrackFindingProcessor.cc have incorrect chi2 and mva variables. The chi2rz and chi2rphi should not be pdof and the mva should be in the 0-1 range and not binned upon tttrack initialization.

tomalin commented on L1Trigger/TrackFindingTracklet/interface/TrackMultiplexer.h:
```diff
@@ -12,74 +12,58 @@
-                 tt::StreamsTrack& acceptedTracks,
-                 tt::StreamsStub& lostStubs,
-                 tt::StreamsTrack& lostTracks);
+    void produce(tt::StreamsTrack& streamsTrack, tt::StreamsStub& streamsStub);
```
Given that consumes, produces & produce are key functions in CMSSW, common to all EDProducers, with definate meaning, I'm not keen on seeing the same function names being used in code that is not an EDProducer. e.g. If one searches the code for "produce" to find where EDProducts are being created, one will get false matches in TrackMultiplexer etc.

tomalin replied:
Any progress on this comment?

tschuh replied:
I like using those names, I would recommend to let cmssw reviewer decide.

tomalin commented on L1Trigger/TrackFindingTracklet/src/DuplicateRemoval.cc:
```diff
@@ -0,0 +1,142 @@
+        input_.push_back(nullptr);
+        continue;
+      }
+      vector<Stub*> stubs;
```
Between L48-L78, it's not at all clear what is going on. To save people having to understand every line of code here, please add comment explaining it. You already had streamTrack & streamStub. What is the purpose of creating new FrameStub objects etc. and storing them in your structs?

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/src/HitPatternHelper.cc:
```diff
@@ -21,13 +21,13 @@ namespace hph {
         dataFormats_(dataFormats),
-        dfcot_(dataFormats_.format(trackerTFP::Variable::cot, trackerTFP::Process::kfin)),
-        dfzT_(dataFormats_.format(trackerTFP::Variable::zT, trackerTFP::Process::kfin)),
+        dfcot_(dataFormats_.format(trackerTFP::Variable::cot, trackerTFP::Process::gp)),
```
Is the HitPatternHelper, which is only being used in the Hybrid (or Tracklet) algo,  being made dependent on the GP, which is only used in the TMTT algo. What is going on here?

tschuh replied:
KF depends on GP, Hybrid uses KF.

tomalin commented:
All your code uses a few classes a lot, such as TTBV, Frame, FrameStub & FrameTrack,  StreamStub & StreamTrack, DataFormats & DataFormat, Setup. But they're only documented in a few sentences in /L1Trigger/TrackerTFP/README.md . I suggest adding a dedicated section to this README, explaining these things and giving examples of common manipulations, such as extracting a float from the digi data.

tomalin commented on L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py:
```diff
@@ -151,6 +149,8 @@
 
 # HYBRID: prompt tracking
 if (L1TRKALGO == 'HYBRID'):
+    process.TrackTriggerSetup.GeometricProcessor.ChosenRofZ = 50.0
```
Why is the Hybrid depending on the TMTT GP? And why is the eta range of the Hybrid being set to 2.4? I seem tp recall the L1TrkNtuple_eff_eta.pdf plot showing non-zero efficiency up to eta 2.5 in the past, although that currently doesn't seem to be the case. Is the eta cut in TrackTriggerSetup used only in the DTC & KF? It's also dodgy changing the default value of parameters for the baseline algo in a cfg.py, as anyone else running this algo will have to copy these parameter mods.

tschuh replied:
performing those changes (restoring the past configuration) is dodgy to begin with, they are just made to pass CI.

tschuh replied:
Please make the CI more reluctant so that we don't need these lines of code.

tomalin replied:
@skinnari do you remember if the Hybrid eta range for track reco was 2.5 or 2.4? It seems to be 2.4 now, but I recall it being 2.5.

tomalin commented:
Please improve the comments at the top of the various .h and _cfg.py files with "Demonstrator" in the title, so that anyone wishing to use this code to check SW vs FW would understand what the various modules and doing and how they need to set up Vivado etc. to work with them.
e.g.  TrackerTFP/test/AnalyzerDemonstrator.cc & TrackerTFP/plugins/ProducerDemonstrator.cc both have the same very short comment, despite the fact that one is an ESProducer and the other is an EDAnalyzer. Please make clear what the difference between them is. e.g. What is the ESProducer for? What is the EDAnalyzer doing?

tomalin commented on L1Trigger/TrackFindingTracklet/python/Demonstrator_cfi.py:
```diff
@@ -3,9 +3,12 @@
 TrackTriggerDemonstrator_params = cms.PSet (
 
-  LabelIn  = cms.string( "TrackFindingTrackletProducerDRin"    ), #
+  LabelIn  = cms.string( "TrackFindingTrackletProducerTM"    ), #
```
All _cfi.py and cff.py files should have a comment at the top saying what they are for.

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/python/Demonstrator_cfi.py:
```diff
@@ -3,9 +3,12 @@
   LabelOut = cms.string( "TrackFindingTrackletProducerDR"      ), #
-  DirIPBB  = cms.string( "/heplnw039/tschuh/work/proj/DRinDR/" ), # path to ipbb proj area
-  RunTime  = cms.double( 4.5 )                                    # runtime in us
+  DirIPBB  = cms.string( "/heplnw039/tschuh/work/proj/tmdr/" ), # path to ipbb proj area
```
Please add comment making clear that these parameters are for ModelSim runs of FW.

tschuh replied:
done

tomalin commented on L1Trigger/TrackTrigger/python/Setup_cff.py:
```diff
@@ -0,0 +1,4 @@
+import FWCore.ParameterSet.Config as cms
```
This is an important cfg file. It deserves a comment at the top explaining what it represents.

tschuh replied:
done

tomalin commented on L1Trigger/TrackTrigger/python/Setup_cfi.py:
```diff
@@ -0,0 +1,203 @@
+import FWCore.ParameterSet.Config as cms
```
This is an important cfg file. It deserves a comment at the tThis is an important cfg file. It deserves a comment at the top explaining what it represents.op explaining what it represents.

tschuh replied:
done

tomalin commented on L1Trigger/TrackTrigger/test/CleanRelVal_cfg.py:
```diff
@@ -0,0 +1,78 @@
+################################################################################################
+# Run bit-accurate TMTT L1 tracking emulation. 
+#
+# To run execute do
```
Why is this job called CleanRelVal if it runs the TMTT tracking? Also, the comment inside it says it is called test_cfg.py.
In any case. we already have TrackerTFP/test/test_cfg.py to run the TMTT tracking, so why add this second job?

tschuh replied:
will be obsolete soon

tomalin replied:
When will it be obsolete?

tschuh replied:
already is.

tomalin commented:
Expand existing comment at start of LayerEncoding.h, to explain what the encoding actually is. i.e. How are the layers encoded?

tomalin commented on L1Trigger/TrackerTFP/interface/TrackFindingProcessor.h:
```diff
@@ -0,0 +1,88 @@
+
+namespace trackerTFP {
+
+  // Class to format final tfp output
```
The name TrackFindingProcessor suggests that this class runs the entire TFP chain. But the comment here says that it only does some formatting of the TFP output, (which if true means it should be renamed to something suggesting that). I think it actually makes the TTTracks and the stream objects.

tomalin replied:
Any progress on this?

tschuh replied:
no

tomalin commented on L1Trigger/TrackerTFP/test/AnalyzerDR.cc:
```diff
@@ -32,16 +31,16 @@ using namespace edm;
-  /*! \class  trklet::AnalyzerKFin
-   *  \brief  Class to analyze hardware like structured TTStub Collection generated by Tracklet
+  /*! \class  trackerTFP::AnalyzerDR
+   *  \brief  Class to analyze hardware like structured TTStub Collection generated by Duplicate Removal
```
Since this consumes StreamsStub and StreamsTrack, is it accurate to say it's analyzing TTStubs? And logically, it must be looking at tracks, to check DR performance.

tschuh replied:
done

tomalin commented on L1Trigger/TrackerTFP/test/test_cfg.py:
```diff
@@ -37,24 +37,44 @@
 # specify input MC
 Samples = [
-'/store/mc/CMSSW_12_6_0/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_125X_mcRun4_realistic_v5_2026D88PU200RV183v2-v1/30000/0959f326-3f52-48d8-9fcf-65fc41de4e27.root'
+'/store/relval/CMSSW_14_0_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_133X_mcRun4_realistic_v1_STD_2026D98_PU200_RV229-v1/2580000/0b2b0b0b-f312-48a8-9d46-ccbadc69bbfd.root',
```
Do we really need dozens of root files here? Consider just one as an example. Same comment applies elsewhere.

tschuh replied:
done

tomalin commented on SimTracker/TrackTriggerAssociation/plugins/CleanAssoc.cc:
```diff
@@ -0,0 +1,83 @@
+namespace tt {
+
+  /*! \class  tt::CleanAssoc
+   *  \brief  creates TTClusterAssociationMap with clean TPs
```
This comment describing the class is far too short. e.g. What is a clean TP? And make clear that this reads in the large version of the association map and writes a reduced version of it.
And does this mean we'll now have two copies of the TTClusterAssociationMap in MC samples? When is this class run?

tschuh replied:
This is not run during Production. I run this when making a trimmed copy from grid to my local pc.

tschuh replied:
I guess it is best to remove those files from the PR, they ended up here by accident.

tschuh replied:
will be obsolete soon

tomalin commented:
Some points in PR description are too short to understand. e.g. "code to create clean TP and association maps added". Please explain what the purpose of this is. And provide evidence (e.g. link to talk) that it is useful. I guess it's goal is to save CPU?

tomalin commented on SimTracker/TrackTriggerAssociation/plugins/CleanAssoc.cc:
```diff
@@ -0,0 +1,83 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
```
If the purpose of this class is to produce a reduced version of the Cluster association map, then it needs renaming. The file name should mention "Clus".

tschuh replied:
will be obsolete soon

tomalin commented:
If it's true that this PR contains a 5-param KF fit, please so say in the description, and give link to a talk where you explain how to use it. e.g. Just one just need to set ParameterFit5 = True in Setup_cfi.py?

tschuh commented:
> I added the comments I could in the parts that I am familiar with. Unfortunately there are a lot of files I am unfamiliar with and so I couldn't be of much help there. I am not opposed to changing tracker quality to its own EDProducer if you feel that it is a better setup. Although right now the HYBRID version still uses it as a function and not EDProducer. Can we make that consistent?
> 
> There are three main concerns I brought up:
> 
>     1. The change in naming scheme from chi2rz and chi2rphi to chi20 and chi21 only adds confusion to any others who look in. I think it's better to change these to the naming scheme that is generally accepted.
> 
>     2. The MVA1 variable is not being calculated correctly as many of the BDT inputs are wrong. The correct variables that should be input can be seen further down in the TrackQuality.cc file and I made notes on changes need to each variable.
> 
>     3. The tttracks being created in TrackFindingProcessor.cc have incorrect chi2 and mva variables. The chi2rz and chi2rphi should not be pdof and the mva should be in the 0-1 range and not binned upon tttrack initialization.

Thanks for your review. I also spotted that there are two mva calculations, one used in the emulator one in the simulator. And those are different... As soon as the new KF replaces the old one the simulation function call will become obsolete and can be removed. to 1) sorry my bad no problem to stick whatever convention you have. to 2. I do not now which emulation reflects the f/w, therefore I do not want to change the behavior. Best person to debug this is @Chriisbrown. to 3. my bad again, I fixed it.

tschuh commented:
I made a forced push squashing all previous commits to make it less painful when I have to rebase in future. I also used the opportunity to remove files which shouldn't be in the PR to begin with.

cgsavard commented on L1Trigger/TrackerTFP/src/TrackFindingProcessor.cc:
```diff
@@ -0,0 +1,264 @@
+      const double ad0 = ttTrackRef->d0();
+      const double aChi2xyfit = it->chi2rphi_;
+      const double aChi2zfit = it->chi2rz_;
+      const double trkMVA1 = (TTTrack_TrackWord::tqMVABins[it->mva_] + TTTrack_TrackWord::tqMVABins[it->mva_ + 1]) / 2.;
```
Won't this crash when the mva_ bin number is already the last one? You can use just ```TTTrack_TrackWord::tqMVABins[it->mva_]``` instead of the bin center as the GTT objects using this variable already assume the bin edge.

tschuh replied:
you are right, I did not expect that something commented as // Bin edges for chi2/dof does not include all bin edges. Fixed as proposed.

tschuh commented:
@tomalin the scram b code-format on the ci machine seems to be different as on my machine. I can't fix the CI failure. Could you please have a look?

tschuh commented:
@tomalin another I realized is that ci complains about L1Trigger/TrackFindingTracklet/interface/Settings.h failing format, however L1Trigger/TrackFindingTracklet/interface/Settings.h gets not touched in this PR ...

cgsavard commented:
I am seeing some pretty big differences in the tracks being produced before and after this big restructuring. For example, the tracks have less nstub, as we can see in this plot where I test 100 evts from ttbar 200 PU:
![Screenshot 2024-09-17 at 4 38 47 PM](https://github.com/user-attachments/assets/03493872-7312-42b5-b152-1a7d1ba17f3d)
And very weirdly, the tracks seem to have a pT cutoff at 15 GeV as seen here:
![Screenshot 2024-09-17 at 4 25 41 PM](https://github.com/user-attachments/assets/aa7a4b01-2d55-4b99-9b25-a6d85c429a64)
56e22-5331-4e8b-9833-91d9c8825ee5)

Other differences involve much smaller chi2 variables and mostly all tracks coming from seed 0 after these changes. Is this all understood?

tschuh commented:
For the pt I found a bug (factor 4) in the TTTrack conversion. The seed 0 dominance originates most likely from the DR, tending to keep lower seeds when duplicates are found. The nStub thing likely originates from the move to a single KF worker. You might want to double check by setting [this](https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/python/Producer_cfi.py#L16) to False. For the chi2s I must confess that I might have changed the behavior. I did not fully understand the original code and replaced it with something which emulates what I would write in f/w. I will look deeper into that.

tschuh commented:
The Chi2s between the old calculation and floating point look similar enough, i don't think that the original code was buggy. Do you look at chi2 or chi2 / ndf?

cgsavard commented:
> For the pt I found a bug (factor 4) in the TTTrack conversion. 

This indeed looks like it's fixed.

> The seed 0 dominance originates most likely from the DR, tending to keep lower seeds when duplicates are found. 

Does this mean the DR behavior has been changed? Because I am seeing that before this PR the seed distribution looks very different.
![Screenshot 2024-09-18 at 11 07 43 AM](https://github.com/user-attachments/assets/5ac5cd33-c044-4b3e-bcb2-ce2fd7e53020)

> The nStub thing likely originates from the move to a single KF worker. You might want to double check by setting [this](https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/python/Producer_cfi.py#L16) to False.

I do see that the nstub distribution looks more similar before and after the PR when setting this variable to false. Based on the variable name and description though, it sounds like this is disabling truncation and not changing the number of KF workers. Is this the same thing? And was this truncation not enabled before and that's why we see differences?

> The Chi2s between the old calculation and floating point look similar enough, i don't think that the original code was buggy. Do you look at chi2 or chi2 / ndf?

Here are the distributions of chi2 and chi2/dof before and after this PR. We can see that it has significantly changed so something does still seem to be off.
![Screenshot 2024-09-18 at 11 13 07 AM](https://github.com/user-attachments/assets/5f8ea6fe-1a9a-45ba-a532-0153afe81267)
![Screenshot 2024-09-18 at 11 13 33 AM](https://github.com/user-attachments/assets/1d61b44d-fd54-46a7-9457-5fa750939833)

tschuh commented:
> Does this mean the DR behavior has been changed?

Yes.

tschuh commented:
> I do see that the nstub distribution looks more similar before and after the PR when setting this variable to false. Based on the variable name and description though, it sounds like this is disabling truncation and not changing the number of KF workers. Is this the same thing? And was this truncation not enabled before and that's why we see differences?

less KF workers means more truncation, therefore is disabling truncation a good test to simulate a many worker scenario

tschuh commented:
> Here are the distributions of chi2 and chi2/dof before and after this PR. We can see that it has significantly changed so something does still seem to be off.

Well the KF got debugged. The old code got sometimes the parameter very wrong leading to bad resolutions and high chi2s. I guess we just see now the debugged chi2 plots. A good cross check would be to compare against oldKF.

cgsavard commented:
> Well the KF got debugged. The old code got sometimes the parameter very wrong leading to bad resolutions and high chi2s. I guess we just see now the debugged chi2 plots. A good cross check would be to compare against oldKF.

Interestingly, I see different behaviors comparing chi2rz and chi2rphi with the old and new kf before these changes were implemented. For chi2rz/dof, I see that the current distribution looks like the old KF before changes. For chi2rphi/dof, I see that the current distribution looks like the new KF before changes. If we assume the oldKF distributions are correct, then it seems the issue might be in the chi2rphi calculation specifically. Here are both distributions below:
![Screenshot 2024-09-23 at 11 33 59 AM](https://github.com/user-attachments/assets/d11804dd-69d9-4ca9-9025-f315aee61efc)
![Screenshot 2024-09-23 at 11 34 27 AM](https://github.com/user-attachments/assets/d19af6e5-5934-4fcc-9960-aa00d7204e09)

cgsavard commented:
> The nStub thing likely originates from the move to a single KF worker. You might want to double check by setting [this](https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/python/Producer_cfi.py#L16) to False.

Is the fact that we now have mostly 4 stub tracks acceptable? Don't we want to be producing more 5-6 stub tracks as they tend to be better tracks? I understand you said this is probably because there is only 1 KF worker, but does this indicate that we should be using more than one worker to get better tracks?

tschuh commented:
> If we assume the oldKF distributions are correct, then it seems the issue might be in the chi2rphi calculation specifically

Well the newKF did show improved r-phi parameter resolution. So it is not unexpected to see lower chi2rphi. The difference is quite big though. I wonder if the oldKF is scaling the chi2rphi up, I think it does that when calculating overall chi2. @tomalin do you know what the oldKF stores in the TTTracks?

tschuh commented:
> Is the fact that we now have mostly 4 stub tracks acceptable? Don't we want to be producing more 5-6 stub tracks as they tend to be better tracks? I understand you said this is probably because there is only 1 KF worker, but does this indicate that we should be using more than one worker to get better tracks?

That is a thing wee need to decide. I am not sure if more worker is sensible as long as we are not able to build f/w. It would be good to understand what we are actually gaining by the additional stubs.

tomalin commented:
git CI reports that the HYBRID_NEWKF performance is significantly better than it used to be. If you've given a talk on this, please link it from the description of this PR. If not, please give such as talk. (However, at eta = 0.05, git CI says HYBRID has z0 resolution of 0.10 cm, whereas HYBRID_NEWKF has one of 0.14cm, so performance should still be improved further, even if not required for this PR).

tomalin commented:
> > I understand that the ProducerTM usually orders tracks according to their seed type. This is important for the functioning of the DR, so should be described in the comment at the top of the class header file, which explains what the class does. And at the line of the code that causes this ordering to happen, an additional comment should appear to highlight it. A comment mentioning this inside ProducerDR would also be wise.
> 
> done

The comment at the top of ProducerTM "Transforms format of TBout into that expected by DR input and muxes all channels to 1. Since DR keeps first tracks the mux ordering is important." still gives no info about what the ordering actually is.

tomalin commented:
> All your code uses a few classes a lot, such as TTBV, Frame, FrameStub & FrameTrack, StreamStub & StreamTrack, DataFormats & DataFormat, Setup. But they're only documented in a few sentences in /L1Trigger/TrackerTFP/README.md . I suggest adding a dedicated section to this README, explaining these things and giving examples of common manipulations, such as extracting a float from the digi data.

Any progress on this comment?

tomalin commented:
> Please improve the comments at the top of the various .h and _cfg.py files with "Demonstrator" in the title, so that anyone wishing to use this code to check SW vs FW would understand what the various modules and doing and how they need to set up Vivado etc. to work with them. e.g. TrackerTFP/test/AnalyzerDemonstrator.cc & TrackerTFP/plugins/ProducerDemonstrator.cc both have the same very short comment, despite the fact that one is an ESProducer and the other is an EDAnalyzer. Please make clear what the difference between them is. e.g. What is the ESProducer for? What is the EDAnalyzer doing?

tomalin commented:
> > Please improve the comments at the top of the various .h and _cfg.py files with "Demonstrator" in the title, so that anyone wishing to use this code to check SW vs FW would understand what the various modules and doing and how they need to set up Vivado etc. to work with them. e.g. TrackerTFP/test/AnalyzerDemonstrator.cc & TrackerTFP/plugins/ProducerDemonstrator.cc both have the same very short comment, despite the fact that one is an ESProducer and the other is an EDAnalyzer. Please make clear what the difference between them is. e.g. What is the ESProducer for? What is the EDAnalyzer doing?

tomalin commented:
> > > Please improve the comments at the top of the various .h and _cfg.py files with "Demonstrator" in the title, so that anyone wishing to use this code to check SW vs FW would understand what the various modules and doing and how they need to set up Vivado etc. to work with them. e.g. TrackerTFP/test/AnalyzerDemonstrator.cc & TrackerTFP/plugins/ProducerDemonstrator.cc both have the same very short comment, despite the fact that one is an ESProducer and the other is an EDAnalyzer. Please make clear what the difference between them is. e.g. What is the ESProducer for? What is the EDAnalyzer doing?

Any progress on this?

tomalin commented:
> Expand existing comment at start of LayerEncoding.h, to explain what the encoding actually is. i.e. How are the layers encoded?

Any progress on this?

tomalin commented:
Hi @tschuh & @cgsavard . I've checked the status of the comments I made on this PR. Those that are still to do, I've either added a "Any progress on this?" comment, which causes them to appear again at the end of this PR conversation or I've marked them with a "thumbs down" icon. Those which are resolved, but where this web page provided no "resolved" button, I've marked with a "thumbs up" icon. Those which are not resolved, but where I caused a copy of the comment to appear later on this page by adding "Any progress on this?", I've marked with a "rocket" icon.
All my comments are trivial styles ones (plus one request to give a talk on performance), so Thomas should be able to answer these easily.
Claire's comments look more substantial, so Thomas should please post a message here when he believes that he's addressed all of her comments.

cgsavard commented:
I did a couple of more studies looking at the old KF vs the new with these changes and I've got a few more concerns.

1. It looks like a lot more high eta duplicate tracks are being produced which is a consequence of the changes in this PR. In general, the duplicate rate for the new KF is higher than the old KF but this PR does help reduce the duplicates in low eta. Do es a fix need to be made for the high eta duplicates or is this understood? Plots that show this: 
<img width="400" alt="image" src="https://github.com/user-attachments/assets/5b9a00c4-29b9-4f8e-ac25-ccf1bb56bf3e">
<img width="400" alt="Screenshot 2024-09-27 at 4 50 09 PM" src="https://github.com/user-attachments/assets/5bb64fe6-0649-428a-867d-92bf1d44b506">

2. I think the increase in z0 resolution issue has been known and I do see that this PR helps fix the resolution a bit, but it's still quite a bit higher from the old KF. If I look at the difference between the tp and match trk z0, I actually see a shift off of 0 for the new KF, both before and after this PR. This is most obvious in the tracking particles with eta > 0. Is there a higher order correction to z0 or another way to fix this? It does seem like the track z0 is being poorly reconstructed.
<img width="400" alt="Screenshot 2024-09-27 at 4 45 03 PM" src="https://github.com/user-attachments/assets/ac8a08ce-2b16-4c5a-bdb9-72fb49fb601e">
<img width="400" alt="Screenshot 2024-09-27 at 4 45 59 PM" src="https://github.com/user-attachments/assets/c8469451-1810-4935-8ced-4d3094de6395">

tschuh commented:
> git CI reports that the HYBRID_NEWKF performance is significantly better than it used to be. If you've given a talk on this, please link it from the description of this PR. If not, please give such as talk. (However, at eta = 0.05, git CI says HYBRID has z0 resolution of 0.10 cm, whereas HYBRID_NEWKF has one of 0.14cm, so performance should still be improved further, even if not required for this PR).

done

tschuh commented:
> > All your code uses a few classes a lot, such as TTBV, Frame, FrameStub & FrameTrack, StreamStub & StreamTrack, DataFormats & DataFormat, Setup. But they're only documented in a few sentences in /L1Trigger/TrackerTFP/README.md . I suggest adding a dedicated section to this README, explaining these things and giving examples of common manipulations, such as extracting a float from the digi data.
> 
> Any progress on this comment?

done

tschuh commented:
> > > > Please improve the comments at the top of the various .h and _cfg.py files with "Demonstrator" in the title, so that anyone wishing to use this code to check SW vs FW would understand what the various modules and doing and how they need to set up Vivado etc. to work with them. e.g. TrackerTFP/test/AnalyzerDemonstrator.cc & TrackerTFP/plugins/ProducerDemonstrator.cc both have the same very short comment, despite the fact that one is an ESProducer and the other is an EDAnalyzer. Please make clear what the difference between them is. e.g. What is the ESProducer for? What is the EDAnalyzer doing?
> 
> Any progress on this?

done

tschuh commented:
> > Expand existing comment at start of LayerEncoding.h, to explain what the encoding actually is. i.e. How are the layers encoded?
> 
> Any progress on this?

done

tschuh commented:
> I did a couple of more studies looking at the old KF vs the new with these changes and I've got a few more concerns.

Many thanks for having a deeper look. Both points (high eta duplicates and resolution bias) are likely worth an issue. Would you mind creating them? For point 1) I am wondering if one can solves this simply by increasing the number of comparison modules [here](https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/python/ChannelAssignment_cfi.py#L18)
For the bias I am wondering if we see that in all 4 track parameter or only in z0.

tschuh commented:
> Claire's comments look more substantial, so Thomas should please post a message here when he believes that he's addressed all of her comments.

I think I answered them all. We discovered are a lot of bugs in Chris's TQ emulation code and I am not sure if we want to fix them in this PR.

tschuh commented:
> git CI reports that the HYBRID_NEWKF performance is significantly better than it used to be. If you've given a talk on this, please link it from the description of this PR. If not, please give such as talk. (However, at eta = 0.05, git CI says HYBRID has z0 resolution of 0.10 cm, whereas HYBRID_NEWKF has one of 0.14cm, so performance should still be improved further, even if not required for this PR).

done

tschuh commented:
> > > I understand that the ProducerTM usually orders tracks according to their seed type. This is important for the functioning of the DR, so should be described in the comment at the top of the class header file, which explains what the class does. And at the line of the code that causes this ordering to happen, an additional comment should appear to highlight it. A comment mentioning this inside ProducerDR would also be wise.
> > 
> > 
> > done
> 
> The comment at the top of ProducerTM "Transforms format of TBout into that expected by DR input and muxes all channels to 1. Since DR keeps first tracks the mux ordering is important." still gives no info about what the ordering actually is.

done

cgsavard commented:
> Many thanks for having a deeper look. Both points (high eta duplicates and resolution bias) are likely worth an issue. Would you mind creating them? 

If they are not addressed in this PR then we can create the issues, but I think the high eta duplicates should be addressed here since it is only showing up as a result of the changes made in this PR.

> For the bias I am wondering if we see that in all 4 track parameter or only in z0.

I don't see any obvious shift in phi and pt but I do see a shift in eta as well where the tttrack tends to have a larger eta than the tracking particle it is matched to. Both the eta and z0 shifts are present before and after this PR in the newKF .

cgsavard commented:
> I am wondering if one can solves this simply by increasing the number of comparison modules

The increase in duplicate high eta tracks does not go away when doubling the number of DR comparison modules to 64, which I just tried.

tomalin commented:
Your code uses MessageLogger via edm::LogPrint("L1Trigger/TrackerTFP"). I believe the category has to be a simple string, so a forward slash is not allowed.
e.g. https://gitlab.cern.ch/ejclemen/cmssw_trackfinding_hlsframework/-/blob/master/TrackFindingTrackletHLS/test/L1tracking_cfg.py?ref_type=heads#L34 gives an example of how the level of printout from the "IRProducer" category is controlled. You can see that the category corresponds to the python variable name, and python variable names cant include forward slash. (Also note that whenever the MessageLogger prints the category name, it automatically also prints the name of the EDProducer that called it). Also LogPrint is intended for "Warnings" https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger , so should only be used when something has gone wrong.

tschuh committed:
squash

tschuh committed:
c++20 and merge fixes

tschuh committed:
code-format

tschuh committed:
run scripts update.

tschuh committed:
run script clean up.

tschuh committed:
rebase fix

tschuh committed:
TrackFindingProcessor for hybrid adapted.

tschuh committed:
script fix

tschuh committed:
little fixes.

tomalin commented:
@tschuh as this looks a significant improvement on the previous HYBRID_NEWKF results, I'm inclined to merge this as soon as it passes git CI. However, can you say here if you've managed to solve any of the issues with the HYBRID_NEWKF performance described by @cgsavard in her talk https://indico.cern.ch/event/1467716/contributions/6179940/attachments/2948559/5182327/L1TK_10_16_24_newKF.pdf ?

tschuh committed:
furthrt little fixes

tschuh committed:
TQ outputs altered.

tomalin commented:
I think @cgsavard 's comments here https://github.com/cms-L1TK/cmssw/pull/285#pullrequestreview-2286805173 have not yet been addressed?

tschuh commented:
> I think @cgsavard 's comments here [#285 (review)](https://github.com/cms-L1TK/cmssw/pull/285#pullrequestreview-2286805173) have not yet been addressed?

those got addressed.

tschuh commented:
> @tschuh as this looks a significant improvement on the previous HYBRID_NEWKF results, I'm inclined to merge this as soon as it passes git CI. However, can you say here if you've managed to solve any of the issues with the HYBRID_NEWKF performance described by @cgsavard in her talk https://indico.cern.ch/event/1467716/contributions/6179940/attachments/2948559/5182327/L1TK_10_16_24_newKF.pdf ?

The issues described by her talk seems to correspond to this [issue](https://github.com/cms-L1TK/cmssw/issues/150). I am not a Tracklet expert and not able to solve it.

tomalin commented:
> > @tschuh as this looks a significant improvement on the previous HYBRID_NEWKF results, I'm inclined to merge this as soon as it passes git CI. However, can you say here if you've managed to solve any of the issues with the HYBRID_NEWKF performance described by @cgsavard in her talk https://indico.cern.ch/event/1467716/contributions/6179940/attachments/2948559/5182327/L1TK_10_16_24_newKF.pdf ?
> 
> The issues described by her talk seems to correspond to this [issue](https://github.com/cms-L1TK/cmssw/issues/150). I am not a Tracklet expert and not able to solve it.

I see that this PR changes the python cfg parameter "UseTTStubResiduals" mentioned in [issue](https://github.com/cms-L1TK/cmssw/issues/150) from False to True. So the KF is no longer by default using the digital output of the Tracklet stage, but instead recalculating the residuals itself. I assume your hypothesis is that the recalculated residuals are still wrong because they are using the helix parameters of the Tracklet seeds, and these you believe are wrong. If we looked at the TTTrack collection produced by the Tracklet part of the chain (i.e. by L1FPGATrackProducer.cc), should we see this? e.g. Should we see the bias in z0 reported in Claire's talk?

tschuh commented:
The old KF in the Hybrid simulation is not fed with stub parameters calculated by the MatchProcessor. It is fed by stub collections determined by the TrackBuilder, but the stub parameters are taken from DTC stubs. Therefore you can't see the corruption in the hybrid s/w chain.

cgsavard commented:
I think that the issues surrounding the TQ and the resolution shifts are ok to be left as an issue and not addressed here as they are separate from the motivation of this PR. 

@tschuh Have you looked into the increase in duplicate tracks at high eta that is a result of the changes made in this PR? This is shown in slide 6 [here](https://indico.cern.ch/event/1467716/contributions/6179940/attachments/2948559/5182327/L1TK_10_16_24_newKF.pdf). I do think this should at be looked into and addressed a little before this PR is merged as this is a direct result of the PR, whereas the other issues were issues that existed beforehand.

tschuh commented:
I will make the mux order of seed types programmable in the TrackMultiplexer, that may give us a handle to improve the DR.

tschuh committed:
TM mux order configurable

tschuh commented:
I am not sure if that changed a lot, summary printout sees now difference.
In order to judge if the higher duplicate rate at high eta is a good or bad thing we need to understand how efficiency and fake rate has changed at high eta. It could be that the KF improved at high eta picking stub combinations which are more likely to match TPs. That will decrease fake rate but also increase efficiency and potentially duplicate rate.

tomalin reviewed (approved):
Since this represents a significant improvement in tracking performance over the old KF, I'm merging it.  The tracking performance still has clear evidence of bugs, as shown in @cgsavard talk https://indico.cern.ch/event/1467716/contributions/6179940/attachments/2948559/5182327/L1TK_10_16_24_newKF.pdf . Thomas suspects these are caused by this issue https://github.com/cms-L1TK/cmssw/issues/150 , which needs debugging.  Claire also suspects the TQ is buggy, but that is unrelated to this PR. She says she's OK with merging this PR too.
